aten/src/ATen/core/jit_type.h

```diff
@@ -1866,15 +1869,22 @@ struct getTypePtr_<c10::Dict<K, V>> final {
     return type;
   }
 };
+
 template <class T>
 struct getTypePtr_<at::optional<T>> final {
+  static const auto callSingleton() {
+    static auto inner_type = getTypePtr_<T>::call();
+    // The "per optional<T>" static singleton needs to live in a .cpp file,
+    // otherwise we'll end up with one singleton instance per shared library.
+    return OptionalType::get(inner_type);
+  }
   static const auto& call() {
-    static auto type = TypeFactory::create<OptionalType>(
-        getTypePtr_<T>::call());
+    static auto type = getTypePtr_<at::optional<T>>::callSingleton();
     return type;
```

Comment:
I'm actually not sure why I had to do this. Originally, I just had this code:
```
static const auto& call() {
    static auto inner_type = getTypePtr_<T>::call();
    static auto type = OptionalType::get(inner_type);
    return type;
```

Which I kinda thought would work, since `OptionalType::get` is defined in a cpp file. I'm not totally sure why, but I got it to work by adding an extra layer of indirection (the `callSingleton()` method).

Reply:
I'm not sure why either, so it makes me very suspicious